source/common/callback/callback.h

```diff
@@ -0,0 +1,175 @@
+ *     void onEvent(const Event& e) override { ... handle event ... }
+ *   };
+ *
+ *   class EventProducer {
+ *   public:
+ *     void attach(EventConsumer* consumer) { consumer_ = consumer; }
+ *     void invoke() { consumer_.onEvent(... some event ...); }
+ *   private:
+ *     EventConsumer* consumer_;
+ *   };
+ *
+ * ... or by passing a functional-style callback like:
+ *
+ *   class MyEventConsumer {
+ *   public:
+ *     MyEventConsumer(EventProducer& producer) {
+ *       producer.attach([this]() { ... handle event ... });
+ *     }
+ *   };
+ *
+ *   class EventProducer {
+ *   public:
+ *     void attach(std::function<void(const Event&)> callback) { callback_ = callback; }
+ *     void invoke() { callback_(... some event ...); }
+ *   private:
+ *     std::function<void(const Event&)> callback_;
+ *   };
+ *
+ *
+ * These approaches are equivalent, and they both have the same issue: the event producer
+ * references the event consumer, either directly or via the lambda's captures, but doesn't manage
+ * its lifetime. If the event consumer is destroyed before the event producer calls it, this is a
+ * use-after-free.
+ *
+ * In some cases, it's straightforward enough to implement "cancelation" by allowing an event
+ * consumer to be detached from any event producers it is currently attached to, but that requires
+ * holding references to all such event producers. That may be impractical or, again, unsafe in the
+ * case where the event consumer outlives its producers.
+ *
+ * Caller and Receiver provide some additional safety. A Receiver owns a callback function, and
+ * can produce Callers which function as weak references to it. The receiver's callback function
+ * can only be invoked via its callers. If the receiver is destroyed, invoking its callers has
+ * no effect, so none of the callback's captures can be unsafely dereferenced.
+ *
+ * When implementing this pattern, an event consumer would own a Receiver and an event producer
+ * would own the corresponding Caller. For example:
+ *
+ *   using EventCaller = Callback::Caller<const Event&>;
+ *   using EventReceiver = Callback::Receiver<const Event&>;
+ *
+ *   class MyEventConsumer {
+ *   public:
+ *     MyEventConsumer(EventProducer& producer) {
+ *       producer.attach(receiver_.caller());
+ *     }
+ *   private:
+ *     EventReceiver receiver_([this]() { ... handle event ... });
+ *   };
+ *
+ *   class EventProducer {
+ *   public:
+ *     void attach(EventCaller caller) { caller_ = caller; }
+ *     void invoke() { caller_(... some event ... ); }
+ *   private:
+ *     EventCaller caller_;
+ *   };
+ */
+namespace Callback {
+
+// Forward-declaration for Caller's friend declaration.
+template <typename... Args> class Receiver;
+
+/**
+ * Caller: simple wrapper for a weak_ptr to a callback function. Copyable and movable.
+ */
+template <typename... Args> class Caller {
+public:
+  /**
+   * Default constructor for default / value initialization.
```

Comment:
Why is this needed?

Reply:
To support the use case where you don't have a Caller until somebody hands one to you, as is the case during initialization:

```c++
class Foo {
public:
  void initialize(Caller<> cb) {
    cb_ = std::move(cb);
    // ...
  }
  void initComplete() {
    cb_();
  }
private:
  Caller<> cb_;
}
```